ui/lib/kv/addon/components/page/secret/metadata/version-diff.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+import { kvDataPath } from 'vault/utils/kv-path';
+import { typeOf } from '@ember/utils';
+
+/**
+ * @module KvVersionDiff
+ * This component produces a JSON diff view between 2 secret versions. It uses the library jsondiffpatch.
+ *
+ * @param {string} backend - Backend from the kv/data model.
+ * @param {string} path - Backend from the kv/data model.
+ * @param {array} metadata - The kv/metadata model. It is version agnostic.
+ * @param {array} breadcrumbs - Array to generate breadcrumbs, passed to the page header component.
+ */
+
+export default class KvVersionDiffComponent extends Component {
+  @service store;
+
+  @tracked leftSideVersion = this.args.metadata.currentVersion;
+  @tracked rightSideVersion = this.defaultRightSideVersion;
+  @tracked statesMatch = false;
+  @tracked visualDiff;
+
+  constructor() {
+    super(...arguments);
+    this.createVisualDiff();
+  }
+
+  get defaultRightSideVersion() {
+    // unless the version is destroyed or deleted we return the version prior to the current version.
+    const versionData = this.args.metadata.sortedVersions.find(
+      (version) =>
+        version.destroyed === false && version.deletion_time === '' && version.version != this.leftSideVersion
+    );
+    return versionData ? versionData.version : this.leftSideVersion - 1;
+  }
+
+  async fetchSecretData(version) {
+    // ARG TODO ask if maybe we want to override the adapter peekRecord?
+    version = typeOf(version) === 'string' ? Number(version) : version;
+    const { backend, path } = this.args;
+    // check the store first, avoiding an extra network request if possible.
+    const storeData = await this.store.peekRecord('kv/data', kvDataPath(backend, path, version));
+    const data = storeData ? storeData : await this.store.queryRecord('kv/data', { backend, path, version });
+
+    return data?.secretData;
+  }
+
+  async createVisualDiff() {
+    /* eslint-disable no-undef */
+    const diffpatcher = jsondiffpatch.create({});
+    const leftSideVersionData = await this.fetchSecretData(this.leftSideVersion);
+    const rightSideVersionData = await this.fetchSecretData(this.rightSideVersion);
+    const delta = diffpatcher.diff(rightSideVersionData, leftSideVersionData);
+
+    this.statesMatch = !delta;
+    this.visualDiff = delta
```

Comment:
could this be a getter that updates based on the left and right sides changing?

Reply:
Update: we explored and came to a similar pattern. Going to resolve this for ease of the next round of reviews.